batchproducer/batchproducer.go

```diff
@@ -233,15 +232,22 @@ func (b *batchProducer) Stop() error {
 }
 
 func (b *batchProducer) setRunning(running bool) {
-	b.runningMu.Lock()
-	defer b.runningMu.Unlock()
-	b.running = running
+	var newValue int32
+	if running {
+		newValue = 1
+	} else {
+		newValue = 0
+	}
+
+	oldValue := b.running
+
+	for swapped := false; !swapped; {
```

Comment:
From what I can tell, we only ever call `setRunning` in `run()`, deferring setting it to false at the end, so my question now moves to whether a single batch producer can ever be shared across multiple threads and/or have it's `run()` method called multiple times. Seems like the method we want to guard against being called multiple times is `run()`.

```go
func (b *bp) run() bool {
  if !atomic.CompareAndSwapInt32(&b.running, 0, 1) {
     return false // already running
  }
  defer atomic.StoreInt32(&b.running, 0)

  // do the things
}
```